### PR TITLE
[AMDGPU] Fix GCNRewritePartialRegUses pass: vector regclass is selected instead of scalar.

### PR DESCRIPTION
**Pull request by vpykhtin**
For the following testcase:

undef %1.sub1:**sgpr_96** = COPY undef %0:sgpr_32
%3:vgpr_32 = V_LSHL_ADD_U32_e64 %1.sub1:sgpr_96, 2, undef %2:vgpr_32, implicit $exec

GCNRewritePartialRegUses produced:

%4:**vgpr_32** = COPY undef %1:sgpr_32
dead %2:vgpr_32 = V_LSHL_ADD_U32_e64 %4, 2, undef %3:vgpr_32, implicit $exec

Register class for %4 is incorrect: there should be _sgpr_32_ instead of _vgpr_32_ because the original %1 had scalar regclass. This happens because %4 participate only in two instructions from which:
-    def in COPY has no reglass information from the instruction description.
-    use in V_LSHL_ADD_U32_e64 has _VS_32_ class from the instruction description.

Thus GCNRewritePartialRegUses used only  _VS_32_ class and selected vector subclass as the largest - what it missed here is that it should take into account the regclass for _%1.sub1:sgpr_96_. This patch fixes that, debug output after the fix:

```
Try to rewrite partial reg %0:SGPR_96
  sub1:SGPR_32 ; <- regclass for %1.sub1:sgpr_96
  sub1:SGPR_32 & VS_32 = SGPR_32 ; <- common class for %1.sub1:sgpr_96 and VS_32 from V_LSHL_ADD_U32_e64 opnd
  Shift 32, reg align 32
  sub1:SGPR_32 -> whole reg, num regclasses 1
  Success %0:SGPR_96 -> %4:SGPR_32
```

And here starts another story. All of this deduction of final regclass using regclasses from instruction operands description is really not needed because _TRI->getSubRegisterClass_ for _%1.sub1:sgpr_96_ is already returning the right class - _SGPR_32_ as we see in the debug output above but that haven't been always like that. I did a bisect search and found that _TRI->getSubRegisterClass_ behavior has been fixed by the https://github.com/llvm/llvm-project/pull/67245, thanks to @kosarev. Before his fix getSubRegisterClass returned strange classes like:

`SGPR_128:sub0_sub1 -> SReg_1_with_sub0_and_SReg_1_with_lo16_in_SGPR_LO16`

which isn't even allocatable.

I'm going to cleanup the implementation but I would like to do this in a separate commit. I've tested my patch before the https://github.com/llvm/llvm-project/pull/67245 and it also works.

Note that GCNRewritePartialRegUses pass isn't enabled by default yet.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp [hunk 1]
```diff
@@ -425,29 +414,41 @@ bool GCNRewritePartialRegUses::rewriteReg(Register Reg) const {
...
+        LLVM_DEBUG(dbgs() << TRI->getRegClassName(SubRegRC) << " & "
+                          << TRI->getRegClassName(OpDescRC) << " = ");
+        SubRegRC = TRI->getCommonSubClass(SubRegRC, OpDescRC);
       }
     }
+
+    if (!SubRegRC) {
+      LLVM_DEBUG(dbgs() << "couldn't find target regclass\n");
+      return false;
+    } else
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp:450]: No else after return

**Comment by vpykhtin** [comment]
@arsenm let's submit this and I create another PR to cleanup unnecessary register class deduction, since _TRI->getSubRegisterClass_ works correctly now.

**Comment by vpykhtin** [comment]
Ping.

**Comment by vpykhtin** [comment]
Ping.

**Review by perlfu — COMMENTED** [review]
I think the code change looks fine.
I left a few questions inline about existing code -- you don't necessarily need to address these.

Could you regenerate the tests as a separate NFC commit (or PR), so this PR only shows test changes from the code change?

File: llvm/test/CodeGen/AMDGPU/rewrite-partial-reg-uses.mir [hunk 2]
```diff
@@ -1,32 +1,31 @@
 # NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
 # RUN: llc -mtriple=amdgcn-amd-amdhsa -amdgpu-enable-rewrite-partial-reg-uses=true -verify-machineinstrs -start-before=rename-independent-subregs -stop-after=rewrite-partial-reg-uses %s -o - | FileCheck -check-prefix=CHECK %s
-# RUN: llc -mtriple=amdgcn-amd-amdhsa -amdgpu-enable-rewrite-partial-reg-uses=true -verify-machineinstrs -start-before=rename-independent-subregs %s -o /dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (vpykhtin) [reply llvm/test/CodeGen/AMDGPU/rewrite-partial-reg-uses.mir:3]: Right, the MIR in the test isn't fully valid to be emitted in asm so I deleted this line.

File: llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp [hunk 3]
```diff
@@ -425,29 +414,41 @@ bool GCNRewritePartialRegUses::rewriteReg(Register Reg) const {
...
 
   auto *RC = MRI->getRegClass(Reg);
   LLVM_DEBUG(dbgs() << "Try to rewrite partial reg " << printReg(Reg, TRI)
                     << ':' << TRI->getRegClassName(RC) << '\n');
 
-  // Collect used subregs and constrained reg classes infered from instruction
+  // Collect used subregs and their reg classes infered from instruction
   // operands.
   SubRegMap SubRegs;
   for (MachineOperand &MO : MRI->reg_nodbg_operands(Reg)) {
```
[the remark this answers is not in the document]
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp]: Done.

File: llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp [hunk 4]
```diff
@@ -425,29 +414,41 @@ bool GCNRewritePartialRegUses::rewriteReg(Register Reg) const {
     return false;
 
   for (MachineOperand &MO : Range) {
```
[the remark this answers is not in the document]
Comment (vpykhtin) [reply llvm/lib/Target/AMDGPU/GCNRewritePartialRegUses.cpp]: Yes, it avoids any processing for unsuitable registers. Replaced the loop with any_of.

**Comment by vpykhtin** [comment]
> I think the code change looks fine. I left a few questions inline about existing code -- you don't necessarily need to address these.
> 
> Could you regenerate the tests as a separate NFC commit (or PR), so this PR only shows test changes from the code change?

Thank you for the review, Carl!

I can regenerate the tests but I'm not sure how to do it best here: submit the updated tests, rebase this PR's branch on top of the commit and then force push the branch? Or I can simply merge _main_ with the updated tests commit into this PR's branch?

**Comment by perlfu** [comment]
> I can regenerate the tests but I'm not sure how to do it best here: submit the updated tests, rebase this PR's branch on top of the commit and then force push the branch? Or I can simply merge _main_ with the updated tests commit into this PR's branch?

I do not know what is expected/common practice, I suspect either approach is valid.
I have certainly seen people merge _main_ before, but typically I opt for a rebase and force push.

**Comment by vpykhtin** [comment]
> I do not know what is expected/common practice, I suspect either approach is valid. I have certainly seen people merge _main_ before, but typically I opt for a rebase and force push.

Thanks, done.

**Review by perlfu — APPROVED** [review]
LGTM

**Pull request merged**